File: library/src/main/java/me/shkschneider/skeleton/network/WebService.java
package me.shkschneider.skeleton.network;

import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.Size;
import android.text.TextUtils;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonSyntaxException;

import java.io.BufferedWriter;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Map;

import me.shkschneider.skeleton.data.CharsetHelper;
import me.shkschneider.skeleton.data.FileHelper;
import me.shkschneider.skeleton.data.MimeTypeHelper;
import me.shkschneider.skeleton.helper.LogHelper;
import me.shkschneider.skeleton.java.ClassHelper;
import me.shkschneider.skeleton.java.SkHide;

public class WebService {

    @Deprecated
    public static WebService options(@NonNull final String url) {
        throw new UnsupportedOperationException(Method.OPTIONS.name());
    }

    public static WebService get(@NonNull final String url) {
        return new WebService(Method.GET, url);
    }

    @Deprecated
    public static WebService head(@NonNull final String url) {
        throw new UnsupportedOperationException(Method.HEAD.name());
    }

    public static WebService post(@NonNull final String url) {
        return new WebService(Method.POST, url);
    }

    public static WebService put(@NonNull final String url) {
        return new WebService(Method.PUT, url);
    }

    public static WebService delete(@NonNull final String url) {
        return new WebService(Method.DELETE, url);
    }

    @Deprecated
    public static WebService trace(@NonNull final String url) {
        throw new UnsupportedOperationException(Method.TRACE.name());
    }

    @Deprecated
    public static WebService connect(@NonNull final String url) {
        throw new UnsupportedOperationException(Method.CONNECT.name());
    }

    private final Method mMethod;
    private String mUrl;
    private Class mType;
    private Map<String, String> mHeaders;
    private Map<String, String> mBody;
    private Callback mCallback;

    private WebService(@NonNull final WebService.Method method, @NonNull final String url) {
        mMethod = method;
        mUrl = url;
        mType = JsonElement.class;
    }

    public WebService url(@NonNull final String url) {
        mUrl = url;
        return this;
    }

    @SkHide
    public String url() {
        return mUrl;
    }

    public WebService headers(@Nullable final Map<String, String> headers) {
        mHeaders = headers;
        return this;
    }

    @SkHide
    public Map<String, String> headers() {
        return mHeaders;
    }

    public WebService body(@Nullable final Map<String, String> body) {
        mBody = body;
        return this;
    }

    @SkHide
    public Map<String, String> body() {
        return mBody;
    }

    public <T> WebService callback(@Nullable final WebService.Callback<T> callback) {
        mCallback = callback;
        return this;
    }

    @SkHide
    public Callback callback() {
        return mCallback;
    }

    public WebService as(final Class type) {
        mType = type;
        return this;
    }

    @SkHide
    public Class as() {
        return mType;
    }

    public void run() {
        new Task().execute();
    }

    // <https://www.ietf.org/rfc/rfc2616.txt>
    private enum Method {

        OPTIONS,
        GET,
        HEAD,
        POST,
        PUT,
        DELETE,
        TRACE,
        CONNECT;

        public boolean allowsBody() {
            switch (this) {
                case POST:
                case PUT:
                case DELETE: // Although SHOULD be ignored
                    return true;
                default:
                    return false;
            }
        }

    }

    private class Task extends AsyncTask<Void, Void, Object> {

        private static final int TIMEOUT_CONNECT = 15000;
        private static final int TIMEOUT_READ = 60000;

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
        }

        @Override
        protected Object doInBackground(@Size(0) @Nullable final Void... voids) {
            HttpURLConnection httpURLConnection = null;
            try {
                final URL url = new URL(mUrl);
                httpURLConnection = (HttpURLConnection) url.openConnection();
                httpURLConnection.setConnectTimeout(TIMEOUT_CONNECT);
                httpURLConnection.setReadTimeout(TIMEOUT_READ);
                httpURLConnection.setUseCaches(false);
                httpURLConnection.setDoInput(true);
                if (mMethod.name() == null) {
                    return new WebServiceException(WebServiceException.INTERNAL_ERROR, "Bad method");
                }

                if (mHeaders != null) {
                    for (final Map.Entry<String, String> entry : mHeaders.entrySet()) {
                        httpURLConnection.setRequestProperty(entry.getKey(), entry.getValue());
                    }
                }
                final boolean allowsBody = mMethod.allowsBody();
                httpURLConnection.setRequestMethod(mMethod.name());
                httpURLConnection.setDoOutput(allowsBody);
                if (! allowsBody && mBody != null) {
                    return new WebServiceException(WebServiceException.INTERNAL_ERROR, "Body not allowed");
                }
                if (mBody != null) {
                    httpURLConnection.setRequestProperty("Content-Type", MimeTypeHelper.APPLICATION_FORMURLENCODED);
                    final DataOutputStream dataOutputStream = new DataOutputStream(httpURLConnection.getOutputStream());
                    String params = "";
                    for (final String key : mBody.keySet()) {
                        if (! TextUtils.isEmpty(params)) params += "&";
                        params += key + "=" + UrlHelper.encode(mBody.get(key));
                    }
                    final BufferedWriter bufferedWriter = new BufferedWriter(new OutputStreamWriter(dataOutputStream, CharsetHelper.UTF8));
                    bufferedWriter.write(params);
                    bufferedWriter.flush();
                    bufferedWriter.close();
                    dataOutputStream.close();
                }

                LogHelper.debug("=> " + mMethod.name() + " " + mUrl + " " + (mHeaders != null ? mHeaders.toString() : "{}") + " " + (mBody != null ? mBody.toString() : "{}"));
                final int responseCode = httpURLConnection.getResponseCode();
                final String responseMessage = httpURLConnection.getResponseMessage();
                LogHelper.debug("<= " + responseCode + " " + responseMessage + " " + mUrl);
                final InputStream errorStream = httpURLConnection.getErrorStream();
                if (errorStream != null) {
                    final String body = FileHelper.readString(errorStream);
                    if (! TextUtils.isEmpty(body)) {
                        LogHelper.verbose("<- " + body);
                        return new WebServiceException(responseCode, body);
                    }
                    return new WebServiceException(responseCode, responseMessage);
                }
                final InputStream inputStream = httpURLConnection.getInputStream();
                final String body = FileHelper.readString(inputStream);
                LogHelper.verbose("<- " + body);
                if (TextUtils.isEmpty(body)) {
                    return "";
                }
                else {
                    final Object result = new Gson().fromJson(body, mType);
                    if (result != null) {
                        return result;
                    }
                }
                return new WebServiceException(responseCode, responseMessage);
            }
            catch (final JsonSyntaxException e) {
                LogHelper.wtf(e);
                return new WebServiceException(WebServiceException.INTERNAL_ERROR, ClassHelper.simpleName(e.getClass()));
            }
            catch (final MalformedURLException e) {
                LogHelper.wtf(e);
                return new WebServiceException(WebServiceException.INTERNAL_ERROR, ClassHelper.simpleName(e.getClass()));
            }
            catch (final IOException e) {
                LogHelper.wtf(e);
                return new WebServiceException(WebServiceException.INTERNAL_ERROR, ClassHelper.simpleName(e.getClass()));
            }
            finally {
                if (httpURLConnection != null) {
                    httpURLConnection.disconnect();
                }
            }
        }

        @SuppressWarnings("unchecked")
        @Override
        protected void onPostExecute(@Nullable final Object result) {
            super.onPostExecute(result);
            if (result instanceof WebServiceException) {
                mCallback.failure((WebServiceException) result);
            }
            else if (result == null) {
                mCallback.success(null);
            }
            else if (ClassHelper.canonicalName(result.getClass()).equals(ClassHelper.canonicalName(mType))) {
                mCallback.success(result);
            }
            else {
                mCallback.failure(new WebServiceException(WebServiceException.INTERNAL_ERROR,
                        ClassHelper.simpleName(result.getClass()) + " != " + ClassHelper.simpleName(mType)));
            }
        }

    }

    public interface Callback<T> {

        void success(@Nullable final T result);
        void failure(@NonNull final WebServiceException e);

    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/network/WebService.kt
package me.shkschneider.skeleton.network

import android.annotation.SuppressLint
import android.os.AsyncTask
import android.support.annotation.Size
import com.google.gson.*
import me.shkschneider.skeleton.data.CharsetHelper
import me.shkschneider.skeleton.data.FileHelper
import me.shkschneider.skeleton.data.MimeTypeHelper
import me.shkschneider.skeleton.helper.Logger
import me.shkschneider.skeleton.java.SkHide
import java.io.BufferedWriter
import java.io.DataOutputStream
import java.io.IOException
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL

class WebService {

    private val TIMEOUT_CONNECT = 15000
    private val TIMEOUT_READ = 60000

    private val method: WebService.Method
    private val url: String
    private var headers: Map<String, String>? = null
    private var body: Map<String, String>? = null
    private var callback: Callback? = null

    constructor(method: WebService.Method, url: String) {
        this.method = method
        this.url = url
    }

    @SkHide
    fun url(): String? {
        return url
    }

    fun headers(headers: Map<String, String>?): WebService {
        this.headers = headers
        return this
    }

    @SkHide
    fun headers(): Map<String, String>? {
        return headers
    }

    fun body(body: Map<String, String>?): WebService {
        this.body = body
        return this
    }

    @SkHide
    fun body(): Map<String, String>? {
        return body
    }

    fun callback(callback: WebService.Callback?): WebService {
        this.callback = callback
        return this
    }

    @SkHide
    fun callback(): WebService.Callback? {
        return callback
    }

    fun run() {
        Task().execute()
    }

    // <https://www.ietf.org/rfc/rfc2616.txt>
    enum class Method {

        OPTIONS,
        GET,
        HEAD,
        POST,
        PUT,
        DELETE,
        TRACE,
        CONNECT;

        fun allowsBody(): Boolean {
            return when (this) {
                POST, PUT, DELETE // Although SHOULD be ignored
                -> true
                else -> false
            }
        }

    }

    @SuppressLint("StaticFieldLeak")
    private inner class Task : AsyncTask<Void, Void, Any?>() {

        override fun doInBackground(@Size(0) vararg voids: Void): Any? {
            var httpURLConnection: HttpURLConnection? = null
            try {
                val url = URL(url)
                httpURLConnection = url.openConnection() as HttpURLConnection
                with (httpURLConnection) {
                    connectTimeout = TIMEOUT_CONNECT
                    readTimeout = TIMEOUT_READ
                    useCaches = false
                    doInput = true
                    headers?.forEach {
                        setRequestProperty(it.key, it.value)
                    }
                    requestMethod = method.name
                    doOutput = method.allowsBody()
                    body?.let {
                        if (! doOutput) {
                            return WebServiceException(WebServiceException.INTERNAL_ERROR, "Body not allowed")
                        }
                        setRequestProperty("Content-Type", MimeTypeHelper.APPLICATION_FORMURLENCODED)
                        val dataOutputStream = DataOutputStream(outputStream)
                        var params = ""
                        it.keys.forEach {
                            params += it + "=" + UrlHelper.encode(it)
                        }
                        val bufferedWriter = BufferedWriter(OutputStreamWriter(dataOutputStream, CharsetHelper.UTF8))
                        bufferedWriter.write(params)
                        bufferedWriter.flush()
                        bufferedWriter.close()
                        dataOutputStream.close()
                    }
                    Logger.debug("=> " + method.name + " " + url + " " + (headers?.toString() ?: "{}") + " " + (body?.toString() ?: "{}"))
                    Logger.debug("<= $responseCode $responseMessage $url")
                    return errorStream?.let {
                        FileHelper.readString(errorStream).takeIf {
                            return it?.isNotBlank()
                        } ?.let {
                            Logger.verbose("<- $it")
                            return it
                        } ?: run {
                            return WebServiceException(responseCode, responseMessage)
                        }
                    }
                }
            } catch (e: JsonSyntaxException) {
                Logger.wtf(e)
                return WebServiceException(WebServiceException.INTERNAL_ERROR, JsonSyntaxException::class.java.simpleName)
            } catch (e: MalformedURLException) {
                Logger.wtf(e)
                return WebServiceException(WebServiceException.INTERNAL_ERROR, MalformedURLException::class.java.simpleName)
            } catch (e: IOException) {
                Logger.wtf(e)
                return WebServiceException(WebServiceException.INTERNAL_ERROR, IOException::class.java.simpleName)
            } finally {
                httpURLConnection?.disconnect()
            }
        }

        override fun onPostExecute(result: Any?) {
            super.onPostExecute(result)
            when (result) {
                is WebServiceException -> callback?.failure(result)
                else -> callback?.success(result as? String)
            }
        }

    }

    interface Callback {

        fun success(result: String?)
        fun failure(e: WebServiceException)

    }

}
